presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java

```diff
@@ -99,6 +103,95 @@ public RelationType getRelationType()
         return relation;
     }
 
+    public Optional<AsteriskedIdentifierChainBasis> resolveAsteriskedIdentifierChainBasis(QualifiedName identifierChain)
```

Comment:
There's a lot of repetition for this logic and mutually recursive calls that make the code hard to follow. Here's an idea on how I would structure it:

```java
private boolean findLocally(Predicate<Scope> match)
{
    Scope scope = this;

    do {
        if (match.test(scope)) {
            return true;
        }

        Optional<Scope> parent = scope.getLocalParent();
        if (!parent.isPresent()) {
            break;
        }

        scope = parent.get();
    }
    while (true);

    return false;
}

public Optional<AsteriskedIdentifierChainBasis> resolveAsteriskedIdentifierChainBasis(QualifiedName identifierChain)
{
    int length = identifierChain.getParts().size();

    boolean foundTable = false;
    boolean foundField = false;

    if (length <= 3) {
        foundTable = findLocally(scope -> scope.getRelationType()
                .getAllFields().stream()
                .anyMatch(field -> field.getRelationAlias().isPresent() && field.getRelationAlias().get().equals(identifierChain));
    }

    if (length >= 2) {
        foundField = findLocally(scope -> scope.getRelationType()
                .getAllFields().stream()
                .anyMatch(field1 -> field1.getRelationAlias().isPresent() && field1.getName().isPresent()
                        && field1.getRelationAlias().get().equals(QualifiedName.of(identifierChain.getParts().get(0)))
                        && field1.getName().get().equals(identifierChain.getParts().get(1))
                        && field1.getType() instanceof RowType));
    }

    if (foundTable && foundField) {
        throw new PrestoException(AMBIGUOUS, ...);
    }

    if (foundTable) {
        return Optional.of(TABLE);
    }

    if (foundField) {
        return Optional.of(FIELD);
    }

    return getOuterQueryParent()
            .map(parent -> parent.resolveAsteriskedIdentifierChainBasis(identifierChain))
            .orElse(Optional.empty());
}
```